### PR TITLE
Fix address sanitizer issues from doc tests

### LINKED ISSUE #23148: IndirectILLReductionFWS seg fault on system test
The IndirectILLReductionFWS doc tests has a segmentation fault when running on the build server.

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work.**

Fixes for problems discovered by the address sanitizer during a `docs-test` run in `MantidPlot`.

**Report to:** [nobody]. 

**To test:**

* Code review

Fixes #23148 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/LineProfile.cpp [hunk 1]
```diff
@@ -225,7 +225,8 @@ void profile(std::vector<double> &Xs, std::vector<double> &Ys,
     for (size_t j = limits.widthStart; j < limits.widthEnd; ++j) {
       const size_t iHor = dir == LineDirection::horizontal ? i : j;
       const size_t iVert = dir == LineDirection::horizontal ? j : i;
-      auto iter = ws.histogram(iVert).begin();
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/LineProfile.cpp]: Interesting bug. Can we prevent this by adding an lvalue reference qualifier to `Histogram::begin()` and `end()`, just like I had done for the day-1 methods in `Histogram` such as `setBinEdges()`?

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp [hunk 2]
```diff
@@ -76,10 +76,10 @@ namespace Algorithms {
  */
 MayersSampleCorrectionStrategy::MayersSampleCorrectionStrategy(
     MayersSampleCorrectionStrategy::Parameters params,
-    const Mantid::HistogramData::Histogram &inputHist)
-    : m_pars(params), m_histogram(inputHist), m_tofVals(inputHist.points()),
-      m_histoYSize(inputHist.y().size()), m_muRrange(calculateMuRange()),
-      m_rng(new MersenneTwister(1)) {
+    HistogramData::Histogram inputHist)
+    : m_pars(params), m_histogram(std::move(inputHist)),
+      m_tofVals(m_histogram.points()), m_histoYSize(m_histogram.y().size()),
```
Comment (SimonHeybrock) [inline Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp]: Note that `m_histogram.y().size()` is simply `m_histogram.size()`.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]
Changes look good, just one comment which could prevent the same thing from happening again:

File: Framework/Algorithms/src/LineProfile.cpp (hunk 1 above)
Comment (martyngigg) [reply Framework/Algorithms/src/LineProfile.cpp]: Good idea. :+1:

File: Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp (hunk 2 above)
Comment (martyngigg) [reply Framework/Algorithms/src/SampleCorrections/MayersSampleCorrectionStrategy.cpp]: :+1:

File: Framework/HistogramData/inc/MantidHistogramData/Histogram.h [hunk 3]
```diff
@@ -192,8 +192,12 @@ class MANTID_HISTOGRAMDATA_DLL Histogram {
...
   void convertToFrequencies();
 
-  HistogramIterator begin() const;
-  HistogramIterator end() const;
+  HistogramIterator begin() const &;
+  HistogramIterator end() const &;
+  // Calling begin/end on a temporary is not allowed as the iterator
+  // reference will be immediately invalidated.
+  HistogramIterator begin() const && = delete;
+  HistogramIterator end() const && = delete;
```
Comment (SimonHeybrock) [inline Framework/HistogramData/inc/MantidHistogramData/Histogram.h]: I don't think you actually need to explicitly do that, havinh the `&` should be sufficient. Just like you do not need to delete the non-const overload of a method if there is only a const version?

Comment (martyngigg) [reply Framework/HistogramData/inc/MantidHistogramData/Histogram.h]: I tried that first, put back the original code in `LineProfile` and everything still compiled, which surprised me too. I will double check.

**Review by SimonHeybrock — APPROVED** [review]

**Comment by martyngigg** [comment]
I've rerun the Ubuntu + documentation tests for both Python 2 & 3 a couple times and they have not failed at all. They were failing fairly regularly before so I hope this has fixed those issues.

**Pull request merged**